presto-parquet/src/test/java/com/facebook/presto/parquet/reader/TestEncryption.java

```diff
@@ -0,0 +1,353 @@
+                .withCodec("GZIP")
+                .withPageSize(1000)
+                .withFooterEncryption()
+                .build();
+        decryptAndValidate();
+    }
+
+    @Test
+    public void testMillionRows()
+            throws IOException
+    {
+        MessageType schema = createSchema();
+        String[] encryptColumns = {"name", "gender"};
+        this.inputFile = new EncryptionTestFileBuilder(conf, schema)
+                .withEncryptColumns(encryptColumns)
+                .withNumRecord(1000000)
+                .withCodec("GZIP")
+                .withPageSize(1000)
+                .withFooterEncryption()
+                .build();
+        decryptAndValidate();
+    }
+
+    @Test
+    public void testPlainTextFooter()
+            throws IOException
+    {
+        MessageType schema = createSchema();
+        String[] encryptColumns = {"name", "gender"};
+        this.inputFile = new EncryptionTestFileBuilder(conf, schema)
+                .withEncryptColumns(encryptColumns)
+                .withNumRecord(10000)
+                .withCodec("SNAPPY")
+                .withPageSize(1000)
+                .build();
+        decryptAndValidate();
+    }
+
+    @Test
+    public void testLargePageSize()
+            throws IOException
+    {
+        MessageType schema = createSchema();
+        String[] encryptColumns = {"name", "gender"};
+        this.inputFile = new EncryptionTestFileBuilder(conf, schema)
+                .withEncryptColumns(encryptColumns)
+                .withNumRecord(100000)
+                .withCodec("GZIP")
+                .withPageSize(100000)
+                .withFooterEncryption()
+                .build();
+        decryptAndValidate();
+    }
+
+    @Test
+    public void testAesGcmCtr()
+            throws IOException
+    {
+        MessageType schema = createSchema();
+        String[] encryptColumns = {"name", "gender"};
+        this.inputFile = new EncryptionTestFileBuilder(conf, schema)
+                .withEncryptColumns(encryptColumns)
+                .withNumRecord(100000)
+                .withCodec("GZIP")
+                .withPageSize(1000)
+                .withEncrytionAlgorithm(ParquetCipher.AES_GCM_CTR_V1)
+                .build();
+        decryptAndValidate();
+    }
+
+    private MessageType createSchema()
+    {
+        return new MessageType("schema",
+                new PrimitiveType(OPTIONAL, INT64, "id"),
+                new PrimitiveType(REQUIRED, BINARY, "name"),
+                new PrimitiveType(OPTIONAL, BINARY, "gender"));
+    }
+
+    private void decryptAndValidate()
+            throws IOException
+    {
+        Path path = new Path(inputFile.getFileName());
+        FileSystem fileSystem = path.getFileSystem(conf);
+        FSDataInputStream inputStream = fileSystem.open(path);
+        long fileSize = fileSystem.getFileStatus(path).getLen();
+        Optional<InternalFileDecryptor> fileDecryptor = createFileDecryptor();
```

Comment:
Can fileDecryptor be shared? and looks like no need to make it an Optional?

Reply:
Shareable? yes, made it a member variable. 
Optional? From a test perspective, we don't need it but the signature of readFooter() need optional.

Reply:
Update: It turned out that I am wrong. The decryptor is not sharable. After I share, I got the error: 'ParquetCryptoRuntimeException: Decryptor re-use'.